src/jit/lower.cpp

```diff
@@ -1288,10 +1288,14 @@ void Lowering::LowerArg(GenTreeCall* call, GenTree** ppArg)
     // Assignments/stores at this level are not really placing an argument.
     // They are setting up temporary locals that will later be placed into
     // outgoing regs or stack.
-    if (arg->OperIsStore() || arg->IsArgPlaceHolderNode() || arg->IsNothingNode() || arg->OperIsCopyBlkOp())
+    // Note that atomic ops may be stores and still produce a value.
+    if ((arg->OperIsStore() && !arg->IsValue()) || arg->IsArgPlaceHolderNode() || arg->IsNothingNode() ||
+        arg->OperIsCopyBlkOp())
```

Comment:
Is there a good reason that we couldn't just check `arg->IsValue()`and return whenever it is false?